src/plugins/chart_expressions/expression_metric/public/components/metric_vis.tsx

```diff
@@ -200,14 +213,14 @@ export const MetricVis = ({
   const metricConfigs: MetricSpec['data'][number] = (
     breakdownByColumn ? data.rows : data.rows.slice(0, 1)
   ).map((row, rowIdx) => {
-    const value: number | string =
+    const value: MetricDatum['value'] =
       row[primaryMetricColumn.id] !== null ? row[primaryMetricColumn.id] : NaN;
     const title = breakdownByColumn
       ? formatBreakdownValue(row[breakdownByColumn.id])
       : primaryMetricColumn.name;
     const subtitle = breakdownByColumn ? primaryMetricColumn.name : config.metric.subtitle;
 
-    if (typeof value !== 'number') {
+    if (typeof value === 'string') {
```

Comment:
I see a lot of props repetition here.
Maybe it is possible to refactor into a shared set of props?

```ts
const sharedMetric = {
      title: String(title),
      subtitle,
      icon: config.metric?.icon ? getIcon(config.metric?.icon) : undefined,
      extra: renderSecondaryMetric(data.columns, row, config),
    };

    if (typeof value === 'string') {
      const nonNumericMetric: MetricWText = {
        ...sharedMetric,
        value: formatPrimaryMetric(value),
        color: config.metric.color ?? defaultColor,
      };
      return nonNumericMetric;
    }

    if (isArrayOfStrings(value)) {
      const formattedValues = value.map((v) => formatPrimaryMetric(v));
      const nonNumericArrayMetric: MetricWStringArrayValues = {
        ...sharedMetric,
        value: formattedValues,
        color: config.metric.color ?? defaultColor,
      };
      return nonNumericArrayMetric;
    }

    if (isArrayOfNumbers(value)) {
      const numericArrayMetric: MetricWNumberArrayValues = {
        ...sharedMetric,
        value,
        valueFormatter: formatPrimaryMetric,
        color: config.metric.color ?? defaultColor,
      };
      return numericArrayMetric;
    }

    const baseMetric: MetricWNumber | MetricWNumberArrayValues = {
      ...sharedMetric,
      ...
```

Reply:
Yeah good call, I just duplicated the returns when typescript complained about a union return.

Moved to shared props in 90f6b94